torch/csrc/distributed/rpc/ProcessGroupAgent.cpp

```diff
@@ -53,17 +53,66 @@ Message deserialize(std::istream& is) {
   return Message(std::move(payload), std::move(tensors), type, id);
 }
 
+std::unordered_map<std::string, int> collectNames(
+    const std::string& workerName, std::shared_ptr<c10d::ProcessGroup>& pg) {
+
+  const auto worldSize = pg->getSize();
+  // collect lengths
+  const int64_t length = workerName.length();
+  std::vector<torch::Tensor> inputLength =
+      { torch::tensor({length}, {torch::kInt64}) };
+  std::vector<std::vector<torch::Tensor>> outputLengths (1);
+  for (int i = 0; i < worldSize; ++i) {
+    outputLengths[0].emplace_back(torch::empty({1}, {torch::kInt64}));
+  }
+  pg->allgather(outputLengths, inputLength)->wait();
+
+  // convert collected length tensors into integers
+  std::vector<int64_t> nameLengths(worldSize);
+  int64_t maxLen = 0;
+  for (int i = 0; i < worldSize; ++i) {
+    nameLengths[i] = outputLengths[0][i].storage().data<int64_t>()[0];
+    maxLen = std::max(maxLen, nameLengths[i]);
+  }
+
+  // collect names
+  torch::Tensor nameTensor = torch::empty({maxLen}, torch::kChar);
+  memcpy(nameTensor.storage().data(), workerName.c_str(), length);
+  std::vector<torch::Tensor> inputName = {nameTensor};
+  std::vector<std::vector<torch::Tensor>> outputNames(1);
+  for (int i = 0; i < worldSize; ++i) {
+    outputNames[0].emplace_back(torch::empty({maxLen}, {torch::kChar}));
+  }
+  pg->allgather(outputNames, inputName)->wait();
+
+  // convert collected name tensors into string names
+  std::unordered_map<std::string, int> nameMap(worldSize);
+  for (int i = 0; i < worldSize; ++i) {
+    torch::Tensor& tensor = outputNames[0][i];
+    std::string peerName(
+        (const char*)tensor.storage().data<signed char>(),
+        nameLengths[i]
+    );
+
+    TORCH_CHECK(nameMap.find(peerName) == nameMap.end(),
+        "RpcAgent name ", peerName, " is not unique.");
+
+    nameMap[std::move(peerName)] = i;
+  }
+
+  return nameMap;
+}
+
 } // namespace
 
 ProcessGroupAgent::ProcessGroupAgent(
     std::string workerName,
-    std::unordered_map<std::string, int> nameMap,
     std::shared_ptr<c10d::ProcessGroup> pg)
     : RpcAgent(std::move(workerName), pg->getRank(), processRequestBlocking),
-      nameMap_(std::move(nameMap)),
       stop_(false),
       pg_(std::move(pg)),
       nextId_(0) {
+  nameMap_ = collectNames(workerName_, pg_);
```

Comment:
nit: can move this into the constructor initializer list above.

Reply:
good point, let me do that.